pkg/nvmf/fabrics.go

```diff
@@ -115,24 +168,55 @@ func disconnectSubsys(nqn, ctrl string) (res bool) {
 	return true
 }
 
-func disconnectByNqn(nqn string) int {
+func disconnectByNqn(nqn, hostnqn string) int {
 	ret := 0
 	if len(nqn) > NVMF_NQN_SIZE {
 		klog.Errorf("Disconnect: nqn %s is too long ", nqn)
 		return -EINVAL
 	}
 
+	// delete hostnqn file
+	hostnqnPath := strings.Join([]string{RUN_NVMF, nqn, b64.StdEncoding.EncodeToString([]byte(hostnqn))}, "/")
+	os.Remove(hostnqnPath)
+
 	devices, err := ioutil.ReadDir(SYS_NVMF)
 	if err != nil {
 		klog.Errorf("Disconnect: readdir %s err: %s", SYS_NVMF, err)
 		return -ENOENT
 	}
 
 	for _, device := range devices {
-		if disconnectSubsys(nqn, device.Name()) {
+		if disconnectSubsysWithHostNqn(nqn, hostnqn, device.Name()) {
 			ret++
 		}
 	}
+
+	nqnPath := strings.Join([]string{RUN_NVMF, nqn}, "/")
+	hostnqns, err := ioutil.ReadDir(nqnPath)
+	if err != nil {
+		klog.Errorf("Disconnect: readdir %s err: %v", nqnPath, err)
+		return -ENOENT
+	}
+	if len(hostnqns) <= 0 {
```

Comment:
\>= ?

Reply:
If there is no more files for hostnqn, we will delete a directory for subsysnqn. It's fallback scenario when linux kernel has no sysfs hostnqn file supports. So, it's not wrong.

Reply:
If the Linux kernel does not have sysfs hostnqn file support, `disconnectSubsysWithHostNqn` will always return false, and `ret` will be always 0, but the hostnqn file we created will always exist, so `len(hostnqns)` will be greater than 0, and the fallback scenario will not work.

By the way, could we move `disconnectSubsysWithHostNqn` and `disconnectSubsys` into a loop, then when `disconnectSubsysWithHostNqn` failed,  we can execute `disconnectSubsys` to disconnect all connections instead of looping twice?

Reply:
The hostnqn file we created will not always exist, because we delete the hostnqn file at the beginning of disconnectByNqn().

And I removed the looping twice problem through returning explicitly UnsupportedHostnqnError.